src/confluent_kafka/src/confluent_kafka.c

```diff
@@ -816,15 +816,27 @@ static int TopicPartition_clear (TopicPartition *self) {
 		Py_DECREF(self->error);
 		self->error = NULL;
 	}
+	if (self->metadata) {
+                free(self->metadata);
+		self->metadata = NULL;
+	}
 	return 0;
 }
 
 static void TopicPartition_setup (TopicPartition *self, const char *topic,
 				  int partition, long long offset,
+				  const char *metadata,
```

Comment:
This is the biggest change I've done - making metadata a string rather than a binary object (`bytes` or something like that). While the prior request started with string and then changed to a binary object, I reverted it back to a utf-8 encoded string. 

The reasoning behind this is as such:
1. any integration test with the null byte (or for that matter, any byte > 127) were failing. The reason for the failure was coming from outside the python client (I checked that we were encoding the metadata into the PyObject correctly). It _seemed_ like the broker was treating the metadata as a string. 
2. On investigating further, I found that librdkafka code (src/rdkafka_request.c:rd_kafka_buf_write_topic_partitions and the methods it calls) treats the metadata as a c string (calls strlen on it and so on).
3. To further test this I modified "0099-commit_metadata.c" in librdkafka repository to be generic binary data (with \x00 in the middle) rather than an ascii string. The test started failing. 
4. And as a final verification, I looked at OffsetCommit in the kafka protocol. It expects the metadata to be a nullable string. 

So - based on this I've kept the metadata as a C string.

Please tell me if there's something wrong with this and if there's some way to pass binary data correctly.

This addresses most of the comments on the original PR as well.

Reply:
In the OffsetCommit case, metadata is certainly a string. The only concern I can think of is that `TopicPartition` is a general-use data structure, so if in the future there is a separate need for topic partition metadata that needs to be binary, then our data structure isn't appropriate.

We could perhaps consider naming the parameter `string_metadata` to make this future proof. I'm not really into that idea.

I think i'm in favor of reverting again to a byte array, and doing a check and throwing an exception on the Python binding side if the array contains a 0 or is an otherwise invalid string. The downside of this API is it isn't as strong for the only current (and likely future) use case. And that's not good. But it is more future proof, and I think that is the stronger requirement to optimize on.

Reply:
Reverted to the bytes type again, and I throw an exception for null bytes.

Reply:
Don't put any restrictions on the contents of the metadata, keep it binary and let the broker reject it if it is invalid.